File: py/structure/set/hash/HashOfChainAddress.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys

class HashOfChainAddress:
    def __init__(self):
        self.__hash = [self.__HashNode(0, 0) for _ in range(64)]

    def isEmpty(self) -> bool:
        for node in self.__hash:
            if node.next:
                return False
        return True

    def add(self, key: int, value: int) -> None:
        current = self.__hash[key % 64]
        while current.next:
            current = current.next
            if key == current.key:
                current.value = value
                return
        current.next = self.__HashNode(key, value)

    def erase(self, key: int) -> None:
        current = self.__hash[key % 64]
        while current.next:
            if key == current.next.key:
                current.next = current.next.next
                return
            current = current.next

    def set(self, key: int, value: int) -> None:
        current = self.__hash[key % 64].next
        while current and key != current.key:
            current = current.next

        if current:
            current.value = value

    def get(self, key: int) -> int:
        current = self.__hash[key % 64].next
        while current and key != current.key:
            current = current.next

        if current:
            return current.value
        return -sys.maxsize

    def clear(self) -> None:
        for node in self.__hash:
            node.next = None

    class __HashNode:
        def __init__(self, key: int, value: int):
            self.key, self.value, self.next = key, value, None

File: java/structure/set/hash/HashOfChainAddress.java
package structure.set.hash;

public class HashOfChainAddress {
    private HashNode[] _hash = new HashNode[64];

    public HashOfChainAddress() {
        for (int i = 0; i < 64; ++i) {
            _hash[i] = new HashNode(0, 0);
        }
    }

    public boolean isEmpty() {
        for (HashNode node : _hash) {
            if (null != node.next) {
                return false;
            }
        }
        return true;
    }

    public void add(int key, int value) {
        HashNode current = _hash[key % 64];
        while (null != current.next) {
            current = current.next;
            if (key == current.key) {
                current.value = value;
                return;
            }
        }
        current.next = new HashNode(key, value);
    }

    public void erase(int key) {
        HashNode current = _hash[key % 64];
        while (null != current.next) {
            if (key == current.next.key) {
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }
    }

    public void set(int key, int value) {
        HashNode current = _hash[key % 64].next;
        while (null != current && key != current.key) {
            current = current.next;
        }

        if (null != current) {
            current.value = value;
        }
    }

    public int get(int key) {
        HashNode current = _hash[key % 64].next;
        while (null != current && key != current.key) {
            current = current.next;
        }

        if (null != current) {
            return current.value;
        }
        return Integer.MIN_VALUE;
    }

    public void clear() {
        for (HashNode node : _hash) {
            node.next = null;
        }
    }

    private class HashNode {
        public int key = 0;
        public int value = 0;
        public HashNode next = null;

        public HashNode(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }
}
